File: Section_3/TemperatureConverter.java
class TemperatureConverter {
    private int count = 0; // No manual memory allocation

    public double[] convert(double temp, boolean toCelsius) {
        count++;
        double result = toCelsius ? (temp - 32) * 5 / 9 : (temp * 9 / 5) + 32;
        return new double[]{result, count}; // Memory is managed by GC
    }

    public static void main(String[] args) {
        TemperatureConverter converter = new TemperatureConverter();
        double[] result1 = converter.convert(100, true);
        System.out.println("Converted: " + result1[0] + ", Calls: " + (int)result1[1]);

        double[] result2 = converter.convert(0, false);
        System.out.println("Converted: " + result2[0] + ", Calls: " + (int)result2[1]);
    }
}

File: Section_3/temperature_converter.cpp
#include <iostream>

class TemperatureConverter {
private:
    int* count; // Dynamically allocated memory

public:
    TemperatureConverter() {
        count = new int(0); // Allocating memory for count
    }

    ~TemperatureConverter() {
        delete count; // Freeing memory
    }

    std::pair<double, int> convert(double temp, bool toCelsius) {
        (*count)++;
        double result = toCelsius ? (temp - 32) * 5.0 / 9.0 : (temp * 9.0 / 5.0) + 32.0;
        return {result, *count};
    }
};

int main() {
    TemperatureConverter* converter = new TemperatureConverter(); // Allocating object dynamically

    auto result1 = converter->convert(100, true);
    std::cout << "Converted: " << result1.first << ", Calls: " << result1.second << std::endl;

    auto result2 = converter->convert(0, false);
    std::cout << "Converted: " << result2.first << ", Calls: " << result2.second << std::endl;

    delete converter; // Freeing allocated object
    return 0;
}
